backend/dataall/modules/metadata_forms/services/metadata_form_service.py

```diff
@@ -173,21 +213,40 @@ def delete_metadata_form_field(uri, fieldUri):
     @TenantPolicyService.has_tenant_permission(MANAGE_METADATA_FORMS)
     @MetadataFormAccessService.can_perform('UPDATE FIELDS')
     def batch_metadata_form_field_update(uri, data):
+        to_delete = []
+        to_update = []
+        to_create = []
+
+        # validate all inputs first
+        # if even one input is invalid -- decline whole batch
         for item in data:
             if item.get('metadataFormUri') != uri:
                 raise Exception('property metadataFormUri does not match form uri')
-            if 'uri' not in item:
-                MetadataFormService.create_metadata_form_field(uri=uri, data=item)
-            elif item.get('uri') is not None:
-                if item.get('deleted', False):
-                    MetadataFormService.delete_metadata_form_field(uri=uri, fieldUri=item['uri'])
-                else:
-                    MetadataFormService.update_metadata_form_field(uri=uri, fieldUri=item['uri'], data=item)
-        return MetadataFormService.get_metadata_form_fields(uri=uri)
+            if item.get('uri') is None:
+                MetadataFormParamValidationService.validate_create_field_params(item)
+                to_create.append(item)
+            elif not item.get('deleted', False):
+                MetadataFormParamValidationService.validate_update_field_params(uri, item)
+                to_update.append(item)
+            else:
+                to_delete.append(item['uri'])
+
+        # process sorted items
+        for item in to_delete:
+            MetadataFormService.delete_metadata_form_field(uri, item)
+
+        with get_context().db_engine.scoped_session() as session:
+            for item in to_update:
+                MetadataFormRepository.update_metadata_form_field(session, item['uri'], item)
+            for item in to_create:
+                MetadataFormRepository.create_metadata_form_field(session, uri, item)
+
+        return MetadataFormService.get_metadata_form_fields(uri)
 
     @staticmethod
     @TenantPolicyService.has_tenant_permission(MANAGE_METADATA_FORMS)
     @MetadataFormAccessService.can_perform('UPDATE FIELD')
     def update_metadata_form_field(uri, fieldUri, data):
         with get_context().db_engine.scoped_session() as session:
+            MetadataFormParamValidationService.validate_update_field_params(uri, data)
```

Comment:
are we validating twice if we also include validate here and in `batch_metadata_form_field_update`?

Reply:
in part 6 it's revisioned